### ISSUE TITLE
Very slow-autocompletion

### ISSUE THREAD

**Issue by rsuhada** [jorgenschaefer/elpy#1657]
# Summary
I'm getting very slow auto-completion results - essentially every complete on dot takes ~5 seconds to display. 

e.g. after typying `dd.sele` it takes 5+ seconds to see the two candidates (dd.select and dd.select_dtypes).

General autocompletion (not after dot) varies - sometimes it is immediate but sometimes takes a few seconds as well.

Thank you!

# My configuration
## OS
Ubuntu 18.04.3 LTS
Kernel: 5.0.0-27-generic x86_64 bits: 64

## Result of `(elpy-config)`
```
Elpy Configuration

Emacs.............: 26.2
Elpy..............: 1.31.0
Virtualenv........: tf-cpu (/home/rxs/anaconda3/envs/tf-cpu)
Interactive Python: python 3.7.3 (/home/rxs/anaconda3/bin/python)
RPC virtualenv....: elpy-rpc-venv (/home/rxs/.virtualenvs/elpy-rpc-venv)
 Python...........: python 3.7.3 (/home/rxs/.virtualenvs/elpy-rpc-venv/bin/python)
 Jedi.............: 0.15.1
 Rope.............: 0.14.0
 Autopep8.........: 1.4.4
 Yapf.............: 0.28.0
 Black............: 19.3b0
Syntax checker....: Not found (flake8)
```

## Elpy configuration in my init.el
```
(use-package elpy
  ;; usel m-x elpy-config to check/debug pathsA
  :ensure t
  :defer t
  :init
  ;; to get rid of the annoying message
  (add-to-list 'warning-suppress-types '(undo discard-info))
  :config
  (elpy-enable)
  (flymake-mode-off)
  (setq jedi:complete-on-dot t)
  (setq python-indent-offset 4)
  (setq elpy-rpc-backend "jedi")
)
```

**Comment by gopar** [comment]
Hey, it looks like you have the `jedi.el` package installed along side with `elpy`. Can you try disabling `jedi.el` when entering a python file? Elpy uses it's own process of `jedi`, and there's been issues where `elpy` and `jedi` give weird behaviours when used together (like the one you observing).

Once you try that can you give an update?

Thanks!

**Comment by galaunay** [comment]
> I did have an old stranded jedi package in my elpy directory. I did not call it any place in my init.el, though.

I think @gopar was talking about the `jedi.el` package, that you may need to uninstall with `M-x package-delete RET jedi RET`.

**Comment by rsuhada** [comment]
Yes, I understand - there was no `jedi.el` installed. The issue is still there.
Oneway I can get rid off the hang-ups is disabling `company-mode` in the buffer - but I'd like to have the completion...

**Comment by galaunay** [comment]
Ok so there may be a problem indeed.

What is the value of `elpy-rpc-timeout` and `elpy-get-info-from-shell` in your case ?

Could you try disabling company-mode and running `(elpy-rpc-get-completion)` with the cursor after a `.` (like `dd.` in your example) ?
Just to check if the problem comes from company or Elpy.

**Comment by rsuhada** [comment]
Here are the values:
`elpy-rpc-timeo` is 1
`elpy-get-info-from-shell` is nil

`(elpy-rpc-get-completion)` this function doesn't exist. Perhaps you meant something else?

**Comment by rsuhada** [comment]
I've disabled `company-mode` after a short while I start getting the CPU spikes... so it seems to be elpy...

**Comment by galaunay** [comment]
Forgot an "s": `(elpy-rpc-get-completions)`.

There was some issues with jedi spawning a lot of subprocess. does the process eating the CPU look like this ?
```
/home/some-user/.virtualenvs/elpy-rpc-venv/bin/python -W ignore -m elpy.__main__
```
Is there a lot of them ?

**Comment by rsuhada** [comment]
Executing `(elpy-rpc-get-completions)` gives following error:

>  Debugger entered--Lisp error: (error "Timeout during RPC call get_completions from backend")
>    signal(error ("Timeout during RPC call get_completions from backend"))
>    error("Timeout during RPC call %s from backend" "get_completions")

The process looks **exactly** as you write, except I see only one at >90% CPU usage.

**Comment by galaunay** [comment]
So it may be `jedi` having trouble parsing your code.

Does it also happen with smaller pieces of code ? like:
```python
class Foo(object):
  def __init__(self, a, b):
    self.a = a
    self.b = b

bar = Foo(4, 5)
bar.   # <-- cursor here
```

If not, is it possible to have access to the code you are having trouble with ?
Or a minimal example generating the issue ?

**Comment by saitiku** [comment]
> Forgot an "s": `(elpy-rpc-get-completions)`.
> 
> There was some issues with jedi spawning a lot of subprocess. does the process eating the CPU look like this ?
> 
> ```
> /home/some-user/.virtualenvs/elpy-rpc-venv/bin/python -W ignore -m elpy.__main__
> ```
> 
> Is there a lot of them ?

@galaunay I have been having this issue, when elpy.__main__ runs when I try to autocomplete, and I have Jedi. I believe Jedi starts building indexes often. SO, if I have imported large packages like numpy it takes forever and uses CPU heavily. It also leads to emacs hanging. what do you suggest?
